### PR TITLE
add Efficient Exact Optimization (EXO)

### PR DESCRIPTION
**Pull request by haozheji**
This PR adds EXO loss on pair-wise preferences to the DPO trainer.

Paper: https://arxiv.org/pdf/2402.00856

截屏2024-06-14 19 11 57

截屏2024-06-14 19 12 04

### CODE REVIEW

**Review by younesbelkada — APPROVED** [review]
Nice addition !

**Comment by younesbelkada** [comment]
cc @haozheji can you run th estyling checks `make precommit`

**Review by vwxyzjn — APPROVED** [review]
Perhaps move this to the top of the file? Otherwise LGTM

**Review by younesbelkada — COMMENTED** [review]
Thanks ! Can you run make precommit & fix the merge conflicts before merging ;-) ?

**Comment by haozheji** [comment]
@younesbelkada Hi! Is there anything I should do to proceed this PR?

**Comment by kashif** [comment]
@haozheji  so there are some merge conflicts... can you try to fix them?

**Review by kashif — APPROVED** [review]

**Comment by kashif** [comment]
@haozheji  could you also kindly run:

```
pre-commit run --all-files
```

in the root of TRL to fix the formatting

File: trl/trainer/dpo_trainer.py [hunk 1]
```diff
@@ -1054,6 +1054,15 @@ def dpo_loss(
...
             ) / (1 - 2 * self.label_smoothing)
+        elif self.loss_type == "exo_pair":
+            # eqn (16) of the EXO paper: https://arxiv.org/pdf/2402.00856
+            import math
+            if self.label_smoothing == 0:
+                self.label_smoothing = 1e-3
+            losses = (
+                (self.beta * logits).sigmoid() * ((self.beta * logits).logsigmoid() - math.log(1 - self.label_smoothing)) 
+              + (-self.beta * logits).sigmoid() * ((-self.beta * logits).logsigmoid() - math.log(self.label_smoothing))
+            )
```
Comment (kashif) [inline trl/trainer/dpo_trainer.py:1065]:
```suggestion
            losses = (self.beta * logits).sigmoid() * (
                (self.beta * logits).logsigmoid() - math.log(1 - self.label_smoothing)
            ) + (-self.beta * logits).sigmoid() * ((-self.beta * logits).logsigmoid() - math.log(self.label_smoothing))
```

File: trl/trainer/dpo_trainer.py [hunk 2]
```diff
@@ -1054,6 +1054,14 @@ def dpo_loss(
                 -F.logsigmoid(self.beta * logits) * (1 - self.label_smoothing)
                 + F.logsigmoid(-self.beta * logits) * self.label_smoothing
             ) / (1 - 2 * self.label_smoothing)
+        elif self.loss_type == "exo_pair":
+            # eqn (16) of the EXO paper: https://arxiv.org/pdf/2402.00856
+            import math
```
Comment (kashif) [inline trl/trainer/dpo_trainer.py:1059]:
```suggestion
            import math

```

**Comment by haozheji** [comment]
@kashif Thank you so much for the fix!

**Comment by kashif** [comment]
@haozheji  so it seems that tensors do not have `logsigmoid()` methods do you will need to use `F. logsigmoid(...)` instead

File: trl/trainer/dpo_trainer.py [hunk 3]
```diff
@@ -1054,6 +1054,15 @@ def dpo_loss(
...
+
+            if self.label_smoothing == 0:
+                self.label_smoothing = 1e-3
+            losses = (self.beta * logits).sigmoid() * (
+                (self.beta * logits).logsigmoid() - math.log(1 - self.label_smoothing)
+            ) + (-self.beta * logits).sigmoid() * ((-self.beta * logits).logsigmoid() - math.log(self.label_smoothing))
```
Comment (kashif) [inline trl/trainer/dpo_trainer.py:1065]:
```suggestion
                F.logsigmoid(self.beta * logits) - math.log(1 - self.label_smoothing)
            ) + (-self.beta * logits).sigmoid() * (F.logsigmoid(-self.beta * logits) - math.log(self.label_smoothing))
```
